lib/hooks.dart

```diff
@@ -0,0 +1,97 @@
+@JS()
+library hooks;
+
+import 'package:js/js.dart';
+import 'package:react/react.dart';
+
+@JS()
+abstract class React {
+  external static List<dynamic> useState(dynamic value);
+}
+
+/// The return value of [useState].
+///
+/// The current value of the state is available via [value] and
+/// functions to update it are available via [set] and [setTx].
+///
+/// Note are two rules for using Hooks (<https://reactjs.org/docs/hooks-rules.html>):
+///
+/// * Only call Hooks at the top level.
+/// * Only call Hooks from inside a [DartFunctionComponent].
+///
+/// Learn more: <https://reactjs.org/docs/hooks-state.html>.
+class StateHook<T> {
+  /// The first item of the pair returned by [React.useState].
+  T _value;
+
+  /// The second item in the pair returned by [React.useState].
+  void Function(dynamic) _setValue;
+
+  StateHook(T initialValue) {
+    final result = React.useState(initialValue);
+    _value = result[0];
+    _setValue = result[1];
+  }
+
+  StateHook.init(T init()) {
+    final result = React.useState(allowInterop(init));
+    _value = result[0];
+    _setValue = result[1];
+  }
+
+  /// The current value of the state.
+  ///
+  /// See: <https://reactjs.org/docs/hooks-reference.html#usestate>.
+  T get value => _value;
+
+  /// Updates [value] to [newValue].
+  ///
+  /// See: <https://reactjs.org/docs/hooks-state.html#updating-state>.
+  void set(T newValue) => _setValue(newValue);
+
+  /// Updates [value] to the return value of [computeNewValue].
+  ///
+  /// See: <https://reactjs.org/docs/hooks-reference.html#functional-updates>.
+  void setTx(T computeNewValue(T oldValue)) => _setValue(allowInterop(computeNewValue));
+}
+
+/// When called inside a [DartFunctionComponent], adds a local state to the component.
+///
+/// Returns a [StateHook] with [StateHook.value] initialized to [initialValue].
+///
+/// __Example__:
+///
+///     UseStateTestComponent(Map props) {
+///       final count = react.useState(0);
+///
+///       return react.div({}, [
+///         count.value,
+///         react.button({'onClick': (_) => count.set(0)}, ['Reset']),
+///         react.button({'onClick': (_) => count.setTx((prev) => prev + 1)}, ['+']),
```

Comment:
Is this also possible without using `setTx`?

```dart
react.button({'onClick': (_) => count.set(count.value + 1)}, ['+']),
```

Reply:
You're right, I just wrote it that way because that was the example of functional updates in the React docs:
https://reactjs.org/docs/hooks-reference.html#functional-updates
Should I change it to `set` and use a different example for `setTx`?

Reply:
I think so, yeah.

Reply:
But we could also just make a follow-up ticket to improve the examples.

Reply:
Okay, I updated the example here

Reply:
This is actually a case where you would need to use `setTx`.

Let's say that two setState calls were batched together:
```dart
react.button({'onClick': (_) {
  count.setTx((prev) => prev + 1);
  count.setTx((prev) => prev + 1);
}, ['+2']),
```

Clicking on the button would result in the value being incremented by two.

However, without setTx:
```dart
react.button({'onClick': (_) {
  count.set(count.value + 1);
  count.set(count.value + 1);
}, ['+2']),
```
Clicking on the button would result in the value being incremented by one, which is not expected.

Here, it would be best practice to use transactional setState calls.

React [recommends using transactional state updaters where possible](https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous), and it also uses them in [this corresponding example](https://reactjs.org/docs/hooks-reference.html#functional-updates), so I think we should keep `setTx` here